### PR TITLE
fix(runtime): support trailing comments in .editorconfig

### LINKED ISSUE #25372: editorconfig: fails to parse valid comments
### Problem

The ini parser doesn't support comments at the end of lines. For example:

``` ini
root = true ; comment
```

Outputs the following error:

```
Error detected while processing BufReadPost Autocommands for "*":
Error executing lua callback: root must be either "true" or "false"
stack traceback:
	[C]: in function 'error'
	/usr/share/nvim/runtime/lua/editorconfig.lua:14: in function 'assert'
	/usr/share/nvim/runtime/lua/editorconfig.lua:206: in function 'parse'
	/usr/share/nvim/runtime/lua/editorconfig.lua:236: in function 'config'
```

From the [spec](http://docs.editorconfig.org/en/master/editorconfig-format.html):

> Comments may be added to the ends of non-empty lines.

### Steps to reproduce

``` sh
echo "root = true ; comment" > .editorconfig
nvim --clean -u NONE .editorconfig --headless +q
```

### Expected behavior

No error.

### Neovim version (nvim -v)

NVIM v0.10.0-dev

### Vim (not Nvim) behaves the same?

no, nvim doesn't support 

### Operating system/version

Ubuntu 22.04

### Terminal name/version

alacritty 0.12.2

### $TERM environment variable

xterm-256color

### Installation

Nightly PPA

### PR DESCRIPTION
**Pull request by icholy**
Fixes https://github.com/neovim/neovim/issues/25372

### CODE REVIEW

**Comment by clason** [comment]
There's a number of similar asserts for other checks (besides for `root`); I assume they also should support comments?

It would make sense to refactor the `== 'true'  or == 'false'`  into a separate `validate` function and use that everywhere.

**Comment by icholy** [comment]
@clason The comments should never make it to those assertions. I'm stripping them out before parsing the line.

**Comment by gpanders** [comment]
Thanks, the change looks good but can you add a test case to `editorconfig_spec.lua`?

**Comment by icholy** [comment]
@gpanders is it alright if I just add some comments here? https://github.com/neovim/neovim/blob/c67122791d6e2c36cf4a9611edd672b1d6b2fdc2/test/functional/plugin/editorconfig_spec.lua#L20-L94

**Comment by gpanders** [comment]
Yea that's fine, anything that demonstrates that inline comments work.

While you're at it, can you add some full-line comments in there as well?

File: test/functional/plugin/editorconfig_spec.lua [hunk 1]
```diff
@@ -89,6 +89,13 @@ setup(function()
 
     [max_line_length.txt]
     max_line_length = 42
+
+    ; whole line comment
+    # whole line comment
+    [with_comments.txt] ; end of line comment
+    indent_style = space # end of line comment
+    indent_size = 4
+    tab_width = 4
```
Comment (icholy) [inline test/functional/plugin/editorconfig_spec.lua:98]: @gpanders let me know if this is ok.

**Comment by clason** [comment]
Sorry, I mixed this up with the `parse` function below 🤦

**Review by gpanders — APPROVED** [review]

**Comment by T-727** [comment]
At the risk of seeming like I'm obsessed with specs or something:
Wasn't this already resolved in:
https://github.com/neovim/neovim/pull/23301 ?

**Pull request closed without merging**